README.md
# babashka.curl

[![Clojars Project](https://img.shields.io/clojars/v/babashka/babashka.curl.svg)](https://clojars.org/babashka/babashka.curl)

A tiny [curl](https://curl.haxx.se/) wrapper via idiomatic Clojure, inspired by [clj-http](https://github.com/dakrone/clj-http#philosophy), [Ring](https://github.com/ring-clojure/ring) and friends.

## Status

This library is part of [babashka](https://github.com/babashka/babashka/)
but can also be used with JVM Clojure. Check `CHANGES.md` before
upgrading as the API may still undergo some changes. Contributions welcome.

## Usage

``` clojure
(require '[babashka.curl :as curl])
(require '[clojure.java.io :as io]) ;; optional
(require '[cheshire.core :as json]) ;; optional
```

### GET

Simple `GET` request:

``` clojure
(curl/get "https://httpstat.us/200")
;;=> {:status 200, :body "200 OK", :headers { ... }}
```

### Headers

Passing headers:

``` clojure
(def resp (curl/get "https://httpstat.us/200" {:headers {"Accept" "application/json"}}))
(json/parse-string (:body resp)) ;;=> {"code" 200, "description" "OK"}
```

### Query parameters

Query parameters:

``` clojure
(->
  (curl/get "https://postman-echo.com/get" {:query-params {"q" "clojure"}})
  :body
  (json/parse-string true)
  :args)
;;=> {:q "clojure"}
```

### POST

A `POST` request with a `:body`:

``` clojure
(def resp (curl/post "https://postman-echo.com/post" {:body "From Clojure"}))
(json/parse-string (:body resp)) ;;=> {"args" {}, "data" "", ...}
```

Posting a file as a `POST` body:

``` clojure
(:status (curl/post "https://postman-echo.com/post" {:body (io/file "README.md")}))
;; => 200
```

Posting a stream as a `POST` body:

``` clojure
(:status (curl/post "https://postman-echo.com/post" {:body (io/input-stream "README.md")}))
;; => 200
```

Posting form params:

``` clojure
(:status (curl/post "https://postman-echo.com/post" {:form-params {"name" "Michiel"}}))
;; => 200
```

### Basic auth

Basic auth:

``` clojure
(:body (curl/get "https://postman-echo.com/basic-auth" {:basic-auth ["postman" "password"]}))
;; => "{\"authenticated\":true}"
```

### Download binary

Download a binary file as a stream:

``` clojure
(io/copy
  (:body (curl/get "https://github.com/babashka/babashka/raw/master/logo/icon.png"
    {:as :stream}))
  (io/file "icon.png"))
(.length (io/file "icon.png"))
;;=> 7748
```

### Passing through arguments

Passing raw arguments to `curl` can be done with `:raw-args`:

``` clojure
(:status (curl/get "http://www.clojure.org" {:raw-args ["--max-redirs" "0"] :throw false}))
;;=> 301
```

### Unix sockets

Talking to a UNIX socket:

``` clojure
(-> (curl/get "http://localhost/images/json"
              {:raw-args ["--unix-socket"
                          "/var/run/docker.sock"]})
    :body
    (json/parse-string true)
    first
    :RepoTags)
;;=> ["babashka/babashka:0.0.79-SNAPSHOT"]
```

### URL construction

Using the low-level API for fine grained(and safer) URL construction:

``` clojure
(-> (curl/request {:url {:scheme "https"
                         :host   "httpbin.org"
                         :port   443
                         :path   "/get"
                         :query  "q=test"}})
    :body
    (json/parse-string true))
;;=>
{:args {:q "test"},
 :headers
 {:Accept "*/*",
  :Host "httpbin.org",
  :User-Agent "curl/7.64.1",
  :X-Amzn-Trace-Id
  "Root=1-5e63989e-7bd5b1dba75e951a84d61b6a"},
 :origin "46.114.35.45",
 :url "https://httpbin.org/get?q=test"}
```

### Exceptions

An `ExceptionInfo` will be thrown for all HTTP response status codes other than `#{200 201 202 203 204 205 206 207 300 301 302 303 304 307}`
or if `curl` exited with a non-zero exit code. The response map is the exception data.

```clojure
(curl/get "https://httpstat.us/404")
;;=> Execution error (ExceptionInfo) at babashka.curl/request (curl.clj:228).
     status 404

(:status (ex-data *e))
;;=> 404
```

To opt out of an exception being thrown, set `:throw` to false.

```clojure
(:status (curl/get "https://httpstat.us/404" {:throw false}))
;;=> 404
```

If the body is being returned as a stream then exceptions are never thrown and the `:exit` value is wrapped in a `Delay`.

```clojure
(:exit (curl/get "https://httpstat.us/404" {:as :stream}))
;;=> #object[clojure.lang.Delay 0x75769ab0 {:status :pending, :val nil}]
(force *1)
;;=> 0
```

### Error output

Error output can be found under the `:err` key:

``` clojure
(:err (curl/get "httpx://postman-echo.com/get" {:throw false}))
;;=> "curl: (1) Protocol \"httpx\" not supported or disabled in libcurl\n"
```

### Compression

From babashka 0.2.4 onwards, this library will call `curl` with `--compressed`
by default. To opt out, pass `:compressed false` in the options.

### Debugging requests

Set `:debug` to `true` to get debugging information along with the response. The
`:command` value contains the command that was executed to obtain the
response. The `:options` value contains options that were used to construct the
command. Note that all of these values are for debugging only and contain
implementation details that may change in the future.

``` clojure
(def resp (curl/head "https://postman-echo.com/head" {:debug true}))
(:command resp)
;;=> ["curl" "--silent" "--show-error" "--location" "--dump-header" "/var/folders/2m/h3cvrr1x4296p315vbk7m32c0000gp/T/babashka.curl16567082489957878064.headers" "--head" "https://postman-echo.com/head"]
(:options resp)
;;=> {:debug true, :url "https://postman-echo.com/head", :method :head, :header-file #object[java.io.File 0x61d34b4 "/var/folders/2m/h3cvrr1x4296p315vbk7m32c0000gp/T/babashka.curl16567082489957878064.headers"]}
```

## Test

``` clojure
$ clojure -A:test
```

## License

Copyright © 2020 Michiel Borkent

Distributed under the EPL License, same as Clojure. See LICENSE.


CHANGES.md
# Changes

- [#16](https://github.com/babashka/babashka.curl/issues/16): BREAKING!
  Exceptional status codes or nonzero `curl` exit codes will throw exceptions by
  default.

- [#9](https://github.com/babashka/babashka.curl/issues/9): BREAKING! Functions
  like `get`, `post`, etc. now always return a map with `:status`, `:body`, and
  `:headers`.


test/babashka/curl_test.clj
(ns babashka.curl-test
  (:require [babashka.curl :as curl]
            [cheshire.core :as json]
            [clojure.java.io :as io]
            [clojure.string :as str]
            [clojure.test :refer [deftest is testing]])
  (:import (clojure.lang ExceptionInfo)))

(defmethod clojure.test/report :begin-test-var [m]
  (println "===" (-> m :var meta :name))
  (println))

(deftest get-test
  (is (str/includes? (:body (curl/get "https://httpstat.us/200"))
                     "200"))
  (is (= 200
         (-> (curl/get "https://httpstat.us/200"
                       {:headers {"Accept" "application/json"}})
             :body
             (json/parse-string true)
             :code)))
  (testing "query params"
    (is (= {:foo1 "bar1", :foo2 "bar2"}
           (-> (curl/get "https://postman-echo.com/get" {:query-params {"foo1" "bar1" "foo2" "bar2"}})
               :body
               (json/parse-string true)
               :args)))))

(deftest delete-test
  (is (= 200 (:status (curl/delete "https://postman-echo.com/delete")))))

(deftest head-test
  (is (= 200 (:status (curl/head "https://postman-echo.com/head")))))

(deftest post-test
  (is (subs (:body (curl/post "https://postman-echo.com/post"))
            0 10))
  (is (str/includes?
       (:body (curl/post "https://postman-echo.com/post"
                         {:body "From Clojure"}))
       "From Clojure"))
  (testing "file body"
    (is (str/includes?
         (:body (curl/post "https://postman-echo.com/post"
                           {:body (io/file "README.md")}))
         "babashka.curl")))
  (testing "JSON body"
    (let [response (curl/post "https://postman-echo.com/post"
                              {:headers {"Content-Type" "application/json"}
                               :body (json/generate-string {:a "foo"})})
          body (:body response)
          body (json/parse-string body true)
          json (:json body)]
      (is (= {:a "foo"} json))))
  (testing "stream body"
    (is (str/includes?
         (:body (curl/post "https://postman-echo.com/post"
                           {:body (io/input-stream "README.md")}))
         "babashka.curl")))
  (testing "form-params"
    (let [body (:body (curl/post "https://postman-echo.com/post"
                                 {:form-params {"name" "Michiel Borkent"}}))
          body (json/parse-string body true)
          headers (:headers body)
          content-type (:content-type headers)]
      (is (= "application/x-www-form-urlencoded" content-type))))
  ;; TODO:
  #_(testing "multipart"
    (testing "posting file"
      (let [tmp-file (java.io.File/createTempFile "foo" "bar")
            _ (spit tmp-file "Michiel Borkent")
            _ (.deleteOnExit tmp-file)
            body (:body (curl/post "https://postman-echo.com/post"
                                   {:multipart [{:name "file"
                                                 :content (io/file tmp-file)}
                                                {:name "filename" :content (.getPath tmp-file)}
                                                ["file2" (io/file tmp-file)]]}))
            body (json/parse-string body true)
            headers (:headers body)
            content-type (:content-type headers)]
        (is (str/starts-with? content-type "multipart/form-data"))
        (is (:files body))
        (is (str/includes? (-> body :form :filename) "foo"))
        (prn body)))))

(deftest patch-test
  (is (str/includes?
       (:body (curl/patch "https://postman-echo.com/patch"
                          {:body "hello"}))
       "hello")))

(deftest basic-auth-test
  (is (re-find #"authenticated.*true"
               (:body
                (curl/get "https://postman-echo.com/basic-auth"
                          {:basic-auth ["postman" "password"]})))))

(deftest get-response-object-test
  (let [response (curl/get "https://httpstat.us/200")]
    (is (map? response))
    (is (= 200 (:status response)))
    (is (= "200 OK" (:body response)))
    (is (string? (get-in response [:headers "server"]))))

  (testing "response object as stream"
    (let [response (curl/get "https://httpstat.us/200" {:as :stream})]
      (is (map? response))
      (is (= 200 (:status response)))
      (is (instance? java.io.InputStream (:body response)))
      (is (= "200 OK" (slurp (:body response))))))

  (comment
    ;; disabled because of https://github.com/postmanlabs/httpbin/issues/617
    (testing "response object with following redirect"
      (let [response (curl/get "https://httpbin.org/redirect-to?url=https://www.httpbin.org")]
        (is (map? response))
        (is (= 200 (:status response)))))

    (testing "response object without fully following redirects"
      (let [response (curl/get "https://httpbin.org/redirect-to?url=https://www.httpbin.org"
                               {:raw-args ["--max-redirs" "0"]
                                :throw false})]
        (is (map? response))
        (is (= 302 (:status response)))
        (is (= "" (:body response)))
        (is (= "https://www.httpbin.org" (get-in response [:headers "location"])))
        (is (empty? (:redirects response)))))))

(deftest accept-header-test
  (is (= 200
         (-> (curl/get "https://httpstat.us/200"
                       {:accept :json})
             :body
             (json/parse-string true)
             :code))))

(deftest url-encode-query-params-test
  (is (= {"my query param?" "hello there"}
         (-> (curl/get "https://postman-echo.com/get" {:query-params {"my query param?" "hello there"}})
             :body
             (json/parse-string)
             (get "args")))))

(deftest low-level-url-test
  (let [response (-> (curl/request {:url {:scheme "https"
                                          :host   "httpbin.org"
                                          :port   443
                                          :path   "/get"
                                          :query  "q=test"}})
                     :body
                     (json/parse-string true))]
    (is (= {:q "test"} (:args response)))
    (is (= "httpbin.org" (get-in response [:headers :Host])))))

(deftest download-binary-file-as-stream-test
  (testing "download image"
    (let [tmp-file (java.io.File/createTempFile "icon" ".png")]
      (.deleteOnExit tmp-file)
      (io/copy (:body (curl/get "https://github.com/babashka/babashka/raw/master/logo/icon.png" {:as :stream}))
               tmp-file)
      (is (= (.length (io/file "test" "icon.png"))
             (.length tmp-file)))))
  (testing "download image with response headers"
    (let [tmp-file (java.io.File/createTempFile "icon" ".png")]
      (.deleteOnExit tmp-file)
      (let [resp (curl/get "https://github.com/babashka/babashka/raw/master/logo/icon.png" {:as :stream})]
        (is (= 200 (:status resp)))
        (io/copy (:body resp) tmp-file))
      (is (= (.length (io/file "test" "icon.png"))
             (.length tmp-file))))))

(deftest stream-test
  ;; This test aims to test what is tested manually as follows:
  ;; - from https://github.com/enkot/SSE-Fake-Server: npm install sse-fake-server
  ;; - start with: PORT=1668 node fakeserver.js
  ;; - ./bb '(let [resp (curl/get "http://localhost:1668/stream" {:as :stream}) body (:body resp) proc (:process resp)] (prn (take 1 (line-seq (io/reader body)))) (.destroy proc))'
  ;; ("data: Stream Hello!")
  (let [server (java.net.ServerSocket. 1668)
        port (.getLocalPort server)]
    (future (try (with-open
                  [socket (.accept server)
                   out (io/writer (.getOutputStream socket))]
                  (binding [*out* out]
                    (println "HTTP/1.1 200 OK")
                    (println "Content-Type: text/event-stream")
                    (println "Connection: keep-alive")
                    (println)
                    (loop []
                      (try (loop []
                             (println "data: Stream Hello!")
                             (Thread/sleep 20)
                             (recur))
                           (catch Exception _ nil)))))
                 (catch Exception e
                   (prn e))))
    (let [resp (curl/get (str "http://localhost:" port)
                         {:as :stream})
          status (:status resp)
          headers (:headers resp)
          body (:body resp)
          proc (:process resp)]
      (is (= 200 status))
      (is (= "text/event-stream" (get headers "content-type")))
      (is (= (repeat 2 "data: Stream Hello!") (take 2 (line-seq (io/reader body)))))
      (is (= (repeat 10 "data: Stream Hello!") (take 10 (line-seq (io/reader body)))))
      (.destroy proc))))

(deftest command-test
  (let [resp (curl/head "https://postman-echo.com/head" {:debug true})
        command (:command resp)
        opts (:options resp)]
    (is (pos? (.indexOf command "--head")))
    (is (identical? :head (:method opts)))))

(deftest stderr-test
  (testing "should throw"
    (let [ex (is (thrown? ExceptionInfo (curl/get "blah://postman-echo.com/get")))
          ex-data (ex-data ex)]
      (is (contains? ex-data :err))
      (is (str/starts-with? (:err ex-data) "curl: (1)"))
      (is (= 1 (:exit ex-data)))))
  (testing "should not throw"
    (let [resp (curl/get "blah://postman-echo.com/get" {:throw false})]
      (is (contains? resp :err))
      (is (str/starts-with? (:err resp) "curl: (1)"))
      (is (= 1 (:exit resp))))))

(deftest exceptional-status-test
  (testing "should throw"
    (let [ex (is (thrown? ExceptionInfo (curl/get "https://httpstat.us/404")))
          response (ex-data ex)]
      (is (= 404 (:status response)))
      (is (zero? (:exit response)))))
  (testing "should throw when streaming based on status code"
    (let [ex (is (thrown? ExceptionInfo (curl/get "https://httpstat.us/404" {:throw true
                                                                             :as :stream})))
          response (ex-data ex)]
      (is (= 404 (:status response)))
      (is (= "404 Not Found" (slurp (:body response))))
      (is (= "" (slurp (:err response))))
      (is (delay? (:exit response)))
      (is (zero? @(:exit response)))))
  (testing "should not throw"
    (let [response (curl/get "https://httpstat.us/404" {:throw false})]
      (is (= 404 (:status response)))
      (is (zero? (:exit response))))))

(deftest compressed-test
  (is (-> (curl/get "https://api.stackexchange.com/2.2/sites")
          :body (json/parse-string true) :items))
  (is (thrown?
       Exception
       ;; the stackexchange API ALWAYS returns a gzipped response but we ask
       ;; curl to not decompress it
       (-> (curl/get "https://api.stackexchange.com/2.2/sites"
                     {:compressed false})
           :body (json/parse-string true) :items))))


src/babashka/curl.clj
(ns babashka.curl
  (:refer-clojure :exclude [get read-line])
  (:require [clojure.java.io :as io]
            [clojure.string :as str])
  (:import [java.net URLEncoder]
           [java.net URI]
           [java.io File SequenceInputStream ByteArrayInputStream]))

(set! *warn-on-reflection* true)

(def windows? (-> (System/getProperty "os.name")
                  (str/lower-case)
                  (str/includes? "win")))

(def ^:dynamic *defaults*
  {:escape (if windows? #(str/replace % "\"" "\\\"") identity)
   :compressed true
   :throw true})

;;;; Utils

(defn- shell-command
  [args opts]
  (let [args (if windows?
               (mapv (:escape opts) args)
               args)
        pb (ProcessBuilder. ^java.util.List args)
        proc (.start pb)
        _ (when-let [is (:in-stream opts)]
            (with-open [stdin (.getOutputStream proc)]
              (io/copy is stdin)))
        out (.getInputStream proc)
        err (.getErrorStream proc)]
    {:out out
     :err err
     :proc proc}))

(defn- exec-curl [args opts]
  (let [res (shell-command args opts)
        out (:out res)
        err (:err res)
        proc (:proc res)]
    (assoc opts :out out :err err :proc proc)))

(defn- file? [f]
  (when (instance? File f)
    (let [f ^File f]
      (and (.exists f)
           (.isFile f)))))

(defn- input-stream? [x]
  (instance? java.io.InputStream x))

(defn- accept-header [opts]
  (when-let [accept (:accept opts)]
    ["-H" (str "Accept: " (case accept
                            :json "application/json"
                            accept))]))

(defn- url-encode
  "Returns an UTF-8 URL encoded version of the given string."
  [^String unencoded]
  (URLEncoder/encode unencoded "UTF-8"))

(defn- curl-command [opts]
  (let [body (:body opts)
        opts (if body
               (cond-> opts
                 (string? body) (assoc :data-raw body)
                 (file? body) (assoc :in-file body)
                 (input-stream? body) (assoc :in-stream body))
               opts)
        method (when-let [method (:method opts)]
                 (case method
                   :head ["--head"]
                   ["--request" (-> method name str/upper-case)]))
        headers (:headers opts)
        headers (loop [headers* (transient [])
                       kvs (seq headers)]
                  (if kvs
                    (let [[k v] (first kvs)]
                      (recur (reduce conj! headers* ["-H" (str k ": " v)]) (next kvs)))
                    (persistent! headers*)))
        accept-header (accept-header opts)
        form-params (when-let [form-params (:form-params opts)]
                      (loop [params* (transient [])
                             kvs (seq form-params)]
                        (if kvs
                          (let [[k v] (first kvs)
                                v (url-encode v)#_(if (file? v)
                                                    (str "@" (.getPath ^File v))
                                                    (url-encode v))
                                param ["--data" (str (url-encode k) "=" v)]]
                            (recur (reduce conj! params* param) (next kvs)))
                          (persistent! params*))))
        ;; TODO:
        ;; multipart-params (when-let [multipart (:multipart opts)]
        ;;                    (loop [params* (transient [])
        ;;                           xs (seq multipart)]
        ;;                      (if xs
        ;;                        (let [x (first xs)
        ;;                              [k v] (if (vector? x)
        ;;                                      x
        ;;                                      [(or (:part-name x)
        ;;                                           (:name x))
        ;;                                       (:content x)])
        ;;                              v (if (file? v) (str "@" (.getPath ^File v)) v)
        ;;                              param ["--form" (str k "=" v)]]
        ;;                          (recur (reduce conj! params* param) (next xs)))
        ;;                        (persistent! params*))))
        query-params (when-let [qp (:query-params opts)]
                       (loop [params* (transient [])
                              kvs (seq qp)]
                         (if kvs
                           (let [[k v] (first kvs)]
                             (recur (conj! params* (str (url-encode k) "=" (url-encode v))) (next kvs)))
                           (str/join "&" (persistent! params*)))))
        data-raw (:data-raw opts)
        data-raw (when data-raw
                   ["--data-raw" data-raw])
        url (let [url* (:url opts)]
              (cond
                (string? url*)
                url*

                (map? url*)
                (str (URI. ^String (:scheme url*)
                           ^String (:user url*)
                           ^String (:host url*)
                           ^Integer (:port url*)
                           ^String (:path url*)
                           ^String (:query url*)
                           ^String (:fragment url*)))))
        in-file (:in-file opts)
        in-file (when in-file ["-d" (str "@" (.getCanonicalPath ^java.io.File in-file))])
        in-stream (:in-stream opts)
        in-stream (when in-stream ["-d" "@-"])
        basic-auth (:basic-auth opts)
        basic-auth (if (sequential? basic-auth)
                     (str/join ":" basic-auth)
                     basic-auth)
        basic-auth (when basic-auth
                     ["--user" basic-auth])
        header-file (.getPath ^File (:header-file opts))
        stream? (identical? :stream (:as opts))]
    [(conj (reduce into (cond-> ["curl" "--silent" "--show-error" "--location" "--dump-header" header-file]
                          (not (false? (:compressed opts))) (conj "--compressed")
                          ;; tested with SSE server, e.g. https://github.com/enkot/SSE-Fake-Server
                          stream? (conj "-N"))
                   [method headers accept-header data-raw in-file in-stream basic-auth
                    form-params #_multipart-params
                    (:raw-args opts)])
           (str url
                (when query-params
                  (str "?" query-params))))
     opts]))

;;;; End utils

;;;; Response Parsing

(defn- read-headers
  [^File header-file]
  (with-open [^java.io.BufferedReader rdr (io/reader header-file)]
    (loop [lines []]
      (if-let [line (.readLine rdr)]
        (recur (conj lines line))
        lines))))

(defn- read-then-unread
  [^java.io.InputStream is]
  (let [c    (.read is)
        eof? (= -1 c)]
    (if eof?
      is
      (-> (byte-array [c])
          (ByteArrayInputStream.)
          (SequenceInputStream. is)))))

(defn- curl-response->map
  "Parses a curl response input stream into a map"
  [opts]
  (let [is ^java.io.InputStream (:out opts)
        ;; curl does not write to :header-file until stdout is read from once.
        ;; This ensures :status and :headers are parsed when option `:as :stream` is set.
        is (read-then-unread is)
        err (:err opts)
        stream? (identical? :stream (:as opts))
        process (:proc opts)
        [body err exit] (if stream?
                          [is err (delay (.waitFor ^java.lang.Process process))]
                          [(slurp is) (slurp err) (.waitFor ^java.lang.Process process)])
        headers (read-headers (:header-file opts))
        [status headers]
        (reduce (fn [[status parsed-headers :as acc] header-line]
                  (if (str/starts-with? header-line "HTTP/")
                    [(Integer/parseInt (second (str/split header-line  #" "))) parsed-headers]
                    (let [[k v] (str/split header-line #":" 2)]
                      (if (and k v)
                        [status (assoc parsed-headers (str/lower-case k) (str/trim v))]
                        acc))))
                [nil {}]
                headers)
        response {:status status
                  :headers headers
                  :body body
                  :err err
                  :process process
                  :exit exit}]
    response))

;;;; End Response Parsing

(def unexceptional-statuses
  #{200 201 202 203 204 205 206 207 300 301 302 303 304 307})

(defn- should-throw? [response opts]
  (and (:throw opts)
       ;; when streaming, we don't know the exit code yet, so it's too early
       ;; to say if we should throw
       (or (and (not (identical? :stream (:as opts)))
                (not (zero? (:exit response))))
           (let [status (:status response)]
             (and status
                  (not (contains? unexceptional-statuses status)))))))

(defn- build-ex-msg [response]
  (cond
    (:status response)
    (str "babashka.curl: status " (:status response))

    (not (str/blank? (:err response)))
    (:err response)

    :else
    "babashka.curl: error"))

(defn request [opts]
  (let [header-file (File/createTempFile "babashka.curl" ".headers")
        opts (assoc opts :header-file header-file)
        default-opts *defaults*
        opts (merge default-opts opts)
        [args opts] (curl-command opts)
        response (let [response (-> (exec-curl args opts)
                                    (curl-response->map))]
                   (.delete header-file)
                   response)
        response (if (:debug opts)
                   (assoc response
                          :command args
                          :options opts)
                   response)]
    (if (should-throw? response opts)
      (throw (ex-info (build-ex-msg response) response))
      response)))

(defn delete
  ([url] (delete url nil))
  ([url opts]
   (let [opts (assoc opts :url url
                     :method :delete)]
     (request opts))))

(defn head
  ([url] (head url nil))
  ([url opts]
   (let [opts (assoc opts :url url
                     :method :head)]
     (request opts))))

(defn get
  ([url] (get url nil))
  ([url opts]
   (let [opts (assoc opts :url url)]
     (request opts))))

(defn post
  ([url] (post url nil))
  ([url opts]
   (let [opts (assoc opts :url url
                     :method :post)]
     (request opts))))

(defn put
  ([url] (put url nil))
  ([url opts]
   (let [opts (assoc opts :url url
                     :method :put)]
     (request opts))))

(defn patch
  ([url] (patch url nil))
  ([url opts]
   (let [opts (assoc opts :url url
                     :method :patch)]
     (request opts))))

(comment
  ;; after running a python server in the source repo with `python3 -m http.server`
  (request {:url      {:host   "localhost"
                       :scheme "http"
                       :port   8000
                       :path   "/src/babashka"}
            :raw-args ["-L"]})

  (request {:url      {:host   "localhost"
                       :scheme "http"
                       :port   8000
                       :path   "/src/babashka"}
            :response true})

  )
